src/lang/pretty_print.rs

```diff
@@ -0,0 +1,669 @@
+                        .collect::<Vec<_>>()
+                        .join(", ")
+                )
+            }
+            TypedValue::Set(s) => {
+                format!(
+                    "#[{}]",
+                    s.set()
+                        .iter()
+                        .sorted() // Ensure deterministic output
+                        .map(|v| v.to_podlang_string())
+                        .collect::<Vec<_>>()
+                        .join(", ")
+                )
+            }
+            TypedValue::PublicKey(p) => format!("PublicKey({p})"),
+            TypedValue::PodId(p) => {
+                format!("0x{}", p.0.encode_hex::<String>())
+            }
+            TypedValue::Raw(r) => {
+                format!("Raw(0x{})", r.encode_hex::<String>())
+            }
+        }
+    }
+}
+
+fn format_predicate_definition(
+    predicate: &CustomPredicate,
+    indent: usize,
+    batch_context: Option<&CustomPredicateBatch>,
+) -> String {
+    let mut result = String::new();
+    let base_indent = " ".repeat(indent);
+    let statement_indent = " ".repeat(indent + 4);
+
+    format_predicate_signature(&mut result, predicate, &base_indent);
+
+    let conjunction_str = if predicate.conjunction { "AND" } else { "OR" };
+    let _ = writeln!(&mut result, " = {}(", conjunction_str);
+
+    for (i, statement) in predicate.statements.iter().enumerate() {
+        if i > 0 {
+            let _ = writeln!(&mut result);
+        }
+        let _ = write!(
+            &mut result,
+            "{}{}",
+            statement_indent,
+            statement.to_podlang_string_with_batch_context(batch_context)
+        );
+    }
+
+    let _ = write!(&mut result, "\n{})", base_indent);
+    result
+}
+
+fn format_predicate_signature(result: &mut String, predicate: &CustomPredicate, base_indent: &str) {
+    let _ = write!(result, "{}{}", base_indent, predicate.name);
+    let _ = write!(result, "(");
+
+    let mut public_args = predicate
+        .wildcard_names
+        .iter()
+        .take(predicate.args_len)
+        .peekable();
+    while let Some(arg_name) = public_args.next() {
+        let _ = write!(result, "{}", arg_name);
+        if public_args.peek().is_some() {
+            let _ = write!(result, ", ");
+        }
+    }
+
+    let mut private_args = predicate
+        .wildcard_names
+        .iter()
+        .skip(predicate.args_len)
+        .peekable();
+    if private_args.peek().is_some() {
+        if predicate.args_len > 0 {
+            let _ = write!(result, ", ");
+        }
+        let _ = write!(result, "private: ");
+        while let Some(arg_name) = private_args.next() {
+            let _ = write!(result, "{}", arg_name);
+            if private_args.peek().is_some() {
+                let _ = write!(result, ", ");
+            }
+        }
+    }
+
+    let _ = write!(result, ")");
+}
+
+fn format_native_predicate(pred: &NativePredicate) -> &'static str {
```

Comment:
Could you move this to an `impl Display for NativePredicate` in `src/middleware/statement.rs`?
This is useful beyond podlang.

Reply:
Done